src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClientX/IO/TdsWriteStream.cs

```diff
@@ -0,0 +1,274 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Diagnostics;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Data.SqlClient;
+
+namespace Microsoft.Data.SqlClientX.IO
+{
+    /// <summary>
+    /// A stream abstraction over the TDS protocol which works to write the TDS packet
+    /// This stream will take care of wrapping the bytes with the TDS header. 
+    /// </summary>
+    internal class TdsWriteStream : Stream, ITdsWriteStream
+    {
+        private Stream _underlyingStream;
+
+        private byte[] _WriteBuffer;
+
+        // Start at the end of the Tds header
+        internal int WriteBufferOffset { get; private set; } = TdsEnums.HEADER_LEN;
+
+        // This should be set before flushing the buffer.
+        public TdsStreamPacketType PacketHeaderType { get; set; }
+
+        internal byte PacketNumber { get; private set; } = 1; // Packets always start with 1.
+
+        public TdsWriteStream(Stream underLyingStream) : base()
+        {
+            _underlyingStream = underLyingStream;
+            _WriteBuffer = new byte[TdsEnums.DEFAULT_LOGIN_PACKET_SIZE];
```

Comment:
The connection string can specify the packet size from the outset. It might be better to account for that here, to save any future need for reallocations.

Reply:
The Protocol says that the initial packet size is always 4096. The connection string parameter is first negotiated with the server in the Login packet and response and then all the buffers need to be adjusted to follow the new negotiated packet size. 
This implementation reflects the protocol spec for TDS packets.

Reply:
Yes, apologies - just re-checked MS-TDS 2.2.6.4 and you're correct.